app/scripts/controllers/permissions/specifications.js

```diff
@@ -0,0 +1,261 @@
+        validateCaveatAccounts(caveat.value, getIdentities),
+    },
+  };
+};
+
+/**
+ * Gets the specifications for all permissions that will be recognized by the
+ * PermissionController.
+ *
+ * @param {{
+ *   getKeyringAccounts: () => Promise<string[]>,
+ *   getIdentities: () => Record<string, Identity>,
+ * }} options - Options bag.
+ */
+export const getPermissionSpecifications = ({
+  getKeyringAccounts,
+  getIdentities,
+}) => {
+  return {
+    [PermissionKeys.eth_accounts]: {
+      targetKey: PermissionKeys.eth_accounts,
+      allowedCaveats: [CaveatTypes.restrictReturnedAccounts],
+
+      factory: (permissionOptions, requestData) => {
+        if (Array.isArray(permissionOptions.caveats)) {
+          throw new Error(
+            `${
+              PermissionKeys.eth_accounts
+            } error: Received unexpected caveats:\n${JSON.stringify(
+              permissionOptions.caveats,
+              null,
+              2,
+            )}`,
+          );
+        }
+
+        // This value will be further validated as part of the caveat.
+        if (!requestData.approvedAccounts) {
+          throw new Error(
+            `${PermissionKeys.eth_accounts} error: No approved accounts specified.`,
+          );
+        }
+
+        return constructPermission({
+          ...permissionOptions,
+          caveats: [
+            CaveatFactories[CaveatTypes.restrictReturnedAccounts](
+              requestData.approvedAccounts,
+            ),
+          ],
+        });
+      },
+
+      methodImplementation: async (_args) => {
+        const accounts = await getKeyringAccounts();
+        const identities = getIdentities();
+
+        return accounts.sort((firstAddress, secondAddress) => {
+          if (!identities[firstAddress]) {
+            throw new Error(`Missing identity for address: "${firstAddress}".`);
+          } else if (!identities[secondAddress]) {
+            throw new Error(
+              `Missing identity for address: "${secondAddress}".`,
+            );
+          } else if (
+            identities[firstAddress].lastSelected ===
+            identities[secondAddress].lastSelected
+          ) {
+            return 0;
+          } else if (identities[firstAddress].lastSelected === undefined) {
+            return 1;
+          } else if (identities[secondAddress].lastSelected === undefined) {
+            return -1;
+          }
+
+          return (
+            identities[secondAddress].lastSelected -
+            identities[firstAddress].lastSelected
+          );
+        });
+      },
+
+      validator: (permission, _origin, _target) => {
+        const { caveats } = permission;
+        if (
+          !caveats ||
+          caveats.length !== 1 ||
+          caveats[0].type !== CaveatTypes.restrictReturnedAccounts
+        ) {
+          throw new Error(
+            `${
+              PermissionKeys.eth_accounts
+            } error: Invalid caveats:\n${JSON.stringify(caveats, null, 2)}`,
```

Comment:
This error could also be long and could contain ethereum addresses

Reply:
Fixed here: d939d9c05